Refactoring \lorawan --> moving it inside \connectivity.

ashok-rao opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Mbed OS will soon be changing directory structure to the below:

```
connectivity
├── netsocket
├── lorawan
│   ├── mbed_lib.json
│   ├── lorastack
│   ├── source
│   ├── include
│   ├── tests
├── nanostack
├── cellular
├── ...

```

This PR is a part of a wider \connectivity refactoring and implements the above new directory structure for \lorawan.

**Todo:**

~- Move \components\lora ---> \connectivity\lorawan\drivers~
Done as part of commit ```7e2d1ef8```

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
-None-

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
-None-

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
-None-

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Build successes:
  * K64F::GCC_ARM::MBED-BUILD
  * K64F::GCC_ARM::MBED-OS-CONNECTIVITY-LORAWAN-TESTS-TESTS-LORAWAN-LORARADIO

----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------
@LDong-Arm @gpsimenos @rajkan01

Commit:
Mbed OS will soon be changing directory structure to the below:

```
connectivity
├── netsocket
├── lorawan
│   ├── mbed_lib.json // nanostack-interface's mbed_lib.json
│   ├── lorastack
│   ├── tests
├── nanostack
├── cellular
├── ...

```

This PR is a part of a wider \connectivity refactoring and implements the above new directory structure for \lorawan.

Commit:
Adding .mbedignore for tests and modifying greentea test for deprecated API and missing header

Commit:
Moving headers and sources to their respective separate directories based on new component structure

Commit:
Modifying paths in each unit test according to new directory structure

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loramac/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaMac")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loramac/Test_LoRaMac.cpp
+  loramac/Test_LoRaMac.cpp
```
Test in the same directory as `unittest.cmake`:
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaMac.cpp
```
otherwise all paths must be relative to `mbed-os/UNITTESTS`

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loramacchannelplan/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaMacChannelPlan")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loramacchannelplan/Test_LoRaMacChannelPlan.cpp
+  loramacchannelplan/Test_LoRaMacChannelPlan.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaMacChannelPlan.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loramaccommand/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaMacCommand")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loramaccommand/Test_LoRaMacCommand.cpp
+  loramaccommand/Test_LoRaMacCommand.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaMacCommand.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loramaccrypto/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaMacCrypto")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loramaccrypto/Test_LoRaMacCrypto.cpp
+  loramaccrypto/Test_LoRaMacCrypto.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaMacCrypto.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphy/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHY")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphy/Test_LoRaPHY.cpp
+  loraphy/Test_LoRaPHY.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHY.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyas923/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYAS923")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyas923/Test_LoRaPHYAS923.cpp
+  loraphyas923/Test_LoRaPHYAS923.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYAS923.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyau915/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYAU915")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyau915/Test_LoRaPHYAU915.cpp
+  loraphyau915/Test_LoRaPHYAU915.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYAU915.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphycn470/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYCN470")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphycn470/Test_LoRaPHYCN470.cpp
+  loraphycn470/Test_LoRaPHYCN470.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYCN470.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphycn779/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYCN779")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphycn779/Test_LoRaPHYCN779.cpp
+  loraphycn779/Test_LoRaPHYCN779.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYCN779.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyeu433/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYEU433")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyeu433/Test_LoRaPHYEU433.cpp
+  loraphyeu433/Test_LoRaPHYEU433.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYEU433.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyeu868/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYEU868")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyeu868/Test_LoRaPHYEU868.cpp
+  loraphyeu868/Test_LoRaPHYEU868.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYEU868.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyin865/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYIN865")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyin865/Test_LoRaPHYIN865.cpp
+  loraphyin865/Test_LoRaPHYIN865.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYIN865.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphykr920/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYKR920")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphykr920/Test_LoRaPHYKR920.cpp
+  loraphykr920/Test_LoRaPHYKR920.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYKR920.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/loraphyus915/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaPHYUS915")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/loraphyus915/Test_LoRaPHYUS915.cpp
+  loraphyus915/Test_LoRaPHYUS915.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaPHYUS915.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/lorawaninterface/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaWANInterface")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/lorawaninterface/Test_LoRaWANInterface.cpp
+  lorawaninterface/Test_LoRaWANInterface.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaWANInterface.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/lorawanstack/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaWANStack")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/lorawanstack/Test_LoRaWANStack.cpp
+  lorawanstack/Test_LoRaWANStack.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaWANStack.cpp
```

LDong-Arm commented on connectivity/lorawan/tests/UNITTESTS/lorawantimer/unittest.cmake:
```diff
@@ -20,18 +20,18 @@ set(TEST_SUITE_NAME "lorawan_LoRaWANTimer")
 # Test & stub files
 set(unittest-test-sources
-  features/lorawan/lorawantimer/Test_LoRaWANTimer.cpp
+  lorawantimer/Test_LoRaWANTimer.cpp
```
```suggestion
  ${CMAKE_CURRENT_LIST_DIR}/Test_LoRaWANTimer.cpp
```

Commit:
Following 2 changes as part of this commit:

    1. Moving lora drivers from \components\lora to \connectivity\drivers\lora
    2. Incorporating review comments.

LDong-Arm commented on UNITTESTS/CMakeLists.txt:
```diff
@@ -146,11 +146,11 @@ set(unittest-includes-base
-  "${PROJECT_SOURCE_DIR}/../features/lorawan/lorastack/mac"
-  "${PROJECT_SOURCE_DIR}/../features/lorawan/lorastack/phy"
-  "${PROJECT_SOURCE_DIR}/../features/lorawan/system"
+  "${PROJECT_SOURCE_DIR}/../connectivity/lorawan"
```
```suggestion
  "${PROJECT_SOURCE_DIR}/../connectivity
  "${PROJECT_SOURCE_DIR}/../connectivity/lorawan/include/lorawan"
```
Also you need to move `lorawan_types.h` into `include/lorawan`. With those two changes, unit tests compile and pass.

ashok-rao replied:
Done

Commit:
Incorporating review comments

LDong-Arm reviewed (approved):
LGTM

LDong-Arm reviewed (approved):
Still LGTM after the rebase

rajkan01 committed:
Fix UNITTESTS build failure
